### PR TITLE
ASR training with RIR and noise augmentation

### PR DESCRIPTION
**Pull request by jbalam-nv**
- Added a script implementing augmentation with RIR, additive noise and codec transcoding.
- Added changes to record original sampling rate of audio in AudioSegment
- Added AugmentationDataset class to read augmentation files from a tarred audio files
- Some changes to Noise and impulse perturbations in perturb.py

This code needs to be tested with multi-node training, please do not merge.

### CODE REVIEW

**Comment by okuchaiev** [comment]
@jbalam-nv  is this still WIP?

File: examples/asr/experimental/speech_to_text_aug.py
Comment (titu1994) [inline examples/asr/experimental/speech_to_text_aug.py:33]: Even if experimental, these two Perturbation classes should go directly into perturb.py, and not in this class. Model level training scripts should be as concise as possible. And putting them in this class prevents others from using these augmentations without editing or copy pasting code from this script, both of which are problematic.

File: examples/asr/experimental/speech_to_text_aug.py
Comment (titu1994) [inline examples/asr/experimental/speech_to_text_aug.py:190]: Please do not use hard coded Perturbation dicts like this. Define in Yaml, so that users can provide hydra level overrides for file paths (and these can remain the default, but practically they will fail on user environments cause paths won't match)

File: examples/asr/experimental/speech_to_text_aug.py
Comment (titu1994) [inline examples/asr/experimental/speech_to_text_aug.py:232]: @okuchaiev Let's mark this as a an automation goal for the future (with balancing for multi node etc)

File: examples/asr/experimental/speech_to_text_aug.py
Comment (titu1994) [inline examples/asr/experimental/speech_to_text_aug.py:234]: This should be defined entirely in Yaml so users can provide overrides.

File: nemo/collections/asr/parts/perturb.py
Comment (titu1994) [inline nemo/collections/asr/parts/perturb.py:62]: Check if audio dataset is none or not before next() and raise proper error instead of null exceptions.

**Review by titu1994 — CHANGES_REQUESTED** [review]
Overall looks great, need to refactor some dicts into yaml and move some augmentations into perturb.py

**Comment by jbalam-nv** [comment]
@okuchaiev Thanks to you, I am able to run multi-node training now, but still haven't completely done a fine-tuning experiment yet. Should have a result by tomorrow.

File: examples/asr/experimental/speech_to_text_aug.py
Comment (jbalam-nv) [reply examples/asr/experimental/speech_to_text_aug.py:33]: @titu1994 This class is just encapsulates calls to other perturbations. I felt this is not a fundamental perturbation like the other methods we have in perturb.py and that this may be a good example for users to create a custom perturbation. I don't mind moving this to perturb.py, let me know if you still think it should go there.

Comment (titu1994) [reply examples/asr/experimental/speech_to_text_aug.py:33]: The perturb() calls are not particularly trivial, so I believe they should he part of in built augmentations. Ease of use outside a particular script is valuable.

**Comment by jbalam-nv** [comment]
@titu1994 Made the changes you suggested. Now that all the new perturbation methods are in perturb.py there seems to be no need for another speech_to_text.py script, so deleted that and added a new config file that has the new augmentations. I am running a test on Circe with the latest version of this branch and it seems to be going fine so far (1.5 hrs).
Please review the changes when you get a chance.

File: nemo/collections/asr/parts/perturb.py
Comment (titu1994) [inline nemo/collections/asr/parts/perturb.py:797]: You need a brace expander here after this line to properly expand the {} syntax.

**Review by titu1994 — CHANGES_REQUESTED** [review]
Looks mostly good to go, you need a brace expander after the {} replacement but after that it can be merged.

File: nemo/collections/asr/parts/perturb.py
Comment (jbalam-nv) [reply nemo/collections/asr/parts/perturb.py:797]: @titu1994 webdataset takes care of expanding the braces https://github.com/tmbdev/webdataset/blob/244357e68b538f38029ef3b0a5d1464e45aade41/webdataset/dataset.py#L438. I have tested this code and it works. 
We had to do it in TarredAudioToTextDataset because we wanted to assign only a certain number of shards to each gpu. Here we don't really care about it we just want the next noise sample from the dataset and all processes see all the shards.

Comment (titu1994) [reply nemo/collections/asr/parts/perturb.py:797]: Ah sounds good then.

**Review by titu1994 — APPROVED** [review]
LGTM

**Pull request merged**
